File: app/src/main/java/com/designpatterns/demo/behavioral/iterator/java/ConcreteAggregate.java
package com.designpatterns.demo.behavioral.iterator.java;

import java.util.ArrayList;
import java.util.List;

/**
 * 创建具体容器类 实现容器接口中的功能
 *
 * @author zhangzhiguo
 * @version 1.0.0
 * @email: 810903987@qq.com
 * @date 2018/6/26 下午3:15
 * @project_name DesignPatternsDemo
 */
public class ConcreteAggregate implements Aggregate {

    private List<String> list = new ArrayList<>();

    @Override
    public int size() {
        return list.size();
    }

    @Override
    public String get(int location) {
        return list.get(location);
    }

    @Override
    public void add(String tel) {
        list.add(tel);
    }

    @Override
    public void remove(String tel) {
        list.remove(tel);
    }

    @Override
    public Iterator iterator() {
        return new ConcreteIterator(this);
    }
}

File: app/src/main/java/com/designpatterns/demo/behavioral/iterator/kotlin/ConcreteAggregate.kt
package com.designpatterns.demo.behavioral.iterator.kotlin


/**
 *
 * 创建具体容器类 实现容器接口中的功能
 *
 * @author  zhangzhiguo
 * @email: 810903987@qq.com
 * @date 2018/6/26 下午3:38
 * @project_name DesignPatternsDemo
 * @version 1.0.0
 *
 */
class ConcreteAggregate:Aggregate {

    var list=ArrayList<String>()

    override fun size(): Int=list.size

    override fun get(location: Int): String =list[location]

    override fun add(tel: String) {
        list.add(tel)
    }

    override fun remove(tel: String) {
        list.remove(tel)
    }

    override fun iterator(): Iterator {
        return ConcreteIterator(this)
    }
}